File: backend-python/src/api/alchemists.py
from typing import Optional
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from src.database.init import get_db

router = APIRouter()


class AlchemistProfileCreate(BaseModel):
    name: str
    service_start_date: Optional[str] = None


class AlchemistProfileUpdate(BaseModel):
    service_start_date: Optional[str] = None
    profile_image: Optional[str] = None


@router.get("/alchemists")
def list_alchemists():
    """List all alchemist profiles (name and profile_image only)."""
    db = get_db()
    cursor = db.cursor()
    cursor.execute('SELECT name, profile_image FROM alchemist_profiles')
    rows = cursor.fetchall()
    return [dict(row) for row in rows]


@router.get("/alchemist/{name}")
def get_alchemist(name: str):
    """Get a single alchemist profile with potions_completed count."""
    db = get_db()
    cursor = db.cursor()

    cursor.execute('SELECT * FROM alchemist_profiles WHERE name = ?', (name,))
    row = cursor.fetchone()

    if not row:
        raise HTTPException(status_code=404, detail=f"Alchemist '{name}' not found")

    profile = dict(row)

    cursor.execute(
        "SELECT COUNT(*) as count FROM potion_orders WHERE assigned_alchemist = ? AND status = 'Ready for Pickup'",
        (name,)
    )
    count_row = cursor.fetchone()
    profile['potions_completed'] = count_row['count'] if count_row else 0

    return profile


@router.post("/alchemist")
def create_alchemist(data: AlchemistProfileCreate):
    """Create a new alchemist profile."""
    db = get_db()
    cursor = db.cursor()

    try:
        cursor.execute(
            'INSERT INTO alchemist_profiles (name, service_start_date) VALUES (?, ?)',
            (data.name, data.service_start_date)
        )
        db.commit()
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

    cursor.execute('SELECT * FROM alchemist_profiles WHERE name = ?', (data.name,))
    row = cursor.fetchone()
    return dict(row)


@router.put("/alchemist/{name}")
def update_alchemist(name: str, data: AlchemistProfileUpdate):
    """Update an alchemist profile (service_start_date and/or profile_image only)."""
    db = get_db()
    cursor = db.cursor()

    cursor.execute('SELECT * FROM alchemist_profiles WHERE name = ?', (name,))
    row = cursor.fetchone()

    if not row:
        raise HTTPException(status_code=404, detail=f"Alchemist '{name}' not found")

    updates = []
    values = []

    if data.service_start_date is not None:
        updates.append('service_start_date = ?')
        values.append(data.service_start_date)

    if data.profile_image is not None:
        updates.append('profile_image = ?')
        values.append(data.profile_image)

    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")

    values.append(name)
    query = f"UPDATE alchemist_profiles SET {', '.join(updates)} WHERE name = ?"
    cursor.execute(query, values)
    db.commit()

    cursor.execute('SELECT * FROM alchemist_profiles WHERE name = ?', (name,))
    row = cursor.fetchone()
    profile = dict(row)

    cursor.execute(
        "SELECT COUNT(*) as count FROM potion_orders WHERE assigned_alchemist = ? AND status = 'Ready for Pickup'",
        (name,)
    )
    count_row = cursor.fetchone()
    profile['potions_completed'] = count_row['count'] if count_row else 0

    return profile

File: backend/src/api/alchemists.ts
import { Router, Request, Response } from 'express';
import { db } from '../database/init';
import { validateServiceStartDate } from '../utils/validation';

const router = Router();

router.get('/alchemists', (_req: Request, res: Response) => {
  try {
    const rows = db.prepare('SELECT name, profile_image FROM alchemist_profiles').all();
    res.json(rows);
  } catch (err) {
    console.error('Error fetching alchemists:', err);
    res.status(500).json({ error: 'Failed to fetch alchemists' });
  }
});

router.get('/alchemist/:name', (req: Request, res: Response) => {
  const { name } = req.params;

  try {
    const alchemist: any = db
      .prepare('SELECT * FROM alchemist_profiles WHERE name = ?')
      .get(name);

    if (!alchemist) {
      return res.status(404).json({ error: 'Alchemist not found' });
    }

    const countResult: any = db
      .prepare(
        `SELECT COUNT(*) as potions_completed FROM potion_orders
         WHERE assigned_alchemist = ? AND status = 'Ready for Pickup'`
      )
      .get(name);

    res.json({
      ...alchemist,
      potions_completed: countResult.potions_completed
    });
  } catch (err) {
    console.error('Error fetching alchemist:', err);
    res.status(500).json({ error: 'Failed to fetch alchemist' });
  }
});

router.post('/alchemist', (req: Request, res: Response) => {
  const { name, service_start_date, profile_image } = req.body;

  if (!name) {
    return res.status(400).json({ error: 'Name is required' });
  }

  const startDate = service_start_date || new Date().toISOString().split('T')[0];

  const dateError = validateServiceStartDate(startDate);
  if (dateError) {
    return res.status(400).json({ error: dateError });
  }

  try {
    const info = db
      .prepare(
        `INSERT INTO alchemist_profiles (name, service_start_date, profile_image)
         VALUES (?, ?, ?)`
      )
      .run(name, startDate, profile_image || null);

    const row = db
      .prepare('SELECT * FROM alchemist_profiles WHERE id = ?')
      .get(info.lastInsertRowid);

    res.status(201).json(row);
  } catch (err: any) {
    console.error('Error creating alchemist:', err);
    if (typeof err?.message === 'string' && err.message.includes('UNIQUE constraint')) {
      return res.status(409).json({ error: 'Alchemist with this name already exists' });
    }
    res.status(500).json({ error: 'Failed to create alchemist' });
  }
});

router.put('/alchemist/:name', (req: Request, res: Response) => {
  const { name } = req.params;
  const { service_start_date, profile_image } = req.body;

  const updates: string[] = [];
  const values: any[] = [];

  if (service_start_date !== undefined) {
    const dateError = validateServiceStartDate(service_start_date);
    if (dateError) {
      return res.status(400).json({ error: dateError });
    }
    updates.push('service_start_date = ?');
    values.push(service_start_date);
  }
  if (profile_image !== undefined) {
    updates.push('profile_image = ?');
    values.push(profile_image);
  }

  if (updates.length === 0) {
    return res.status(400).json({ error: 'No fields to update' });
  }

  values.push(name);

  try {
    const info = db
      .prepare(`UPDATE alchemist_profiles SET ${updates.join(', ')} WHERE name = ?`)
      .run(...values);

    if (info.changes === 0) {
      return res.status(404).json({ error: 'Alchemist not found' });
    }

    const row: any = db
      .prepare('SELECT * FROM alchemist_profiles WHERE name = ?')
      .get(name);

    const countResult: any = db
      .prepare(
        `SELECT COUNT(*) as potions_completed FROM potion_orders
         WHERE assigned_alchemist = ? AND status = 'Ready for Pickup'`
      )
      .get(name);

    res.json({
      ...row,
      potions_completed: countResult.potions_completed
    });
  } catch (err) {
    console.error('Error updating alchemist:', err);
    res.status(500).json({ error: 'Failed to update alchemist' });
  }
});

export default router;
